Add PERSISTED_FEATURE_FLAGS env

mariusandra opened the pull request:
## Changes

- This adds an env, `PERSIST_FEATURE_FLAGS` that overrides feature flags in the frontend, at least the ones in the `featureFlagLogic`. 
- We can hardcode a value into this env inside settings.py to burn some flags into a release.
- Running with the env `PERSIST_FEATURE_FLAGS=4267-taxonomic-property-filter,4535-funnel-bar-viz` and not having those flags in the system, I had the new filter and the new funnels working.
- All other flags (those not in the env/setting, but still coming in with `/decide`) are passed through as they are.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

mariusandra committed:
add PERSIST_FEATURE_FLAGS env

macobo commented:
Q: What is this solving?

mariusandra commented:
We want to enable code that's behind flags `4267-taxonomic-property-filter` and `4535-funnel-bar-viz` in the next release. We can either make a PR to remove the flags altogether, or hardcode them in, so we could eventually remove them on app and not worry about breaking old deployments running 1.27.

So "what is this solving"? Literally [the issue you opened](https://github.com/PostHog/posthog/issues/2141) with the fix you suggested:  😁 

The idea is that in a release branch before pushing it out, we edit the hardcoded list in settings.py.

mariusandra commented:
Two questions I'd have here:
1. Do we want to force-enable only certain flags like we do now in this PR... and leave everything else dynamic (via `/decide`).... or do we want to completely ignore whatever `/decide` returns.
2. Naming things. `PERSIST`?

timgl commented:
I think this is a decent solution. I wonder if it would be cleaner to before a release just remove all feature flags instead but I don't feel strongly about it.

Do you want to add `4267-taxonomic-property-filter,4535-funnel-bar-viz` to this PR?

mariusandra committed:
improve handling of persisted flags

mariusandra committed:
release three flags

mariusandra committed:
add missing feature flag constants

mariusandra commented:
Actually, yes, let's just really release it. Updates:

- I added `4267-taxonomic-property-filter`, `4535-funnel-bar-viz` and `save-cohort-on-modal` (new one that I found, wasn't in a constant, fixed in https://github.com/PostHog/posthog/pull/5271) to settings.
- If this is merged, then these will now be released on cloud as well to everyone.
- I also found a bunch of non-constant feature flags and moved them to `constants.tsx`. Except for `save-cohort-on-modal` which will be added in that PR.

paolodamico commented on frontend/src/types.ts:
```diff
@@ -974,6 +974,7 @@ export type EventOrPropType = EventDefinition & PropertyDefinition
 export interface AppContext {
     current_user: UserType | null
     preflight: PreflightStatus
+    persist_feature_flags?: string[]
```
Name suggest a boolean, how about `persisted_feature_flags`?

mariusandra replied:
renamed!

paolodamico commented on posthog/settings.py:
```diff
@@ -63,6 +63,14 @@ def get_list(text: str) -> List[str]:
 
+# These flags will be force-enabled on the frontend **in addition to** flags from `/decide`
+# The features here are released, but the flags are just not yet removed from the code.
+PERSIST_FEATURE_FLAGS = get_list(os.getenv("PERSIST_FEATURE_FLAGS", "")) or [
```
I was thinking we could allow a special value, e.g. `False` that allows setting this as an empty list (essentially bypassing this behavior). Alternatively or in addition to we may want to ignore this if `IS_MULTITENANCY`

mariusandra replied:
Setting this env to `"0"` or `"False"` will effectively achieve the same result now, only there will be one unchecked feature flag floating around. I'd say that's good enough to start :)

mariusandra replied:
Regarding cloud... I'd actually leave this on. Otherwise we run the risk of having the self-hosted and cloud environments start to differ by too much...

paolodamico commented on frontend/src/lib/logic/featureFlagLogic.ts:
```diff
@@ -21,17 +22,24 @@ function notifyFlagIfNeeded(flag: string, flagState: boolean): void {
                 get(_, flag) {
                     if (flag === 'toJSON') {
-                        return JSON.stringify(featureFlags)
+                        return () => combinedFlags
```
Context into why this doesn't return the object JSON-stringified?

mariusandra replied:
I made a mistake before. This should return a function that returns a JSON-serializable object
https://dev.to/jankapunkt/how-stringify-proxy-to-json-10oe

paolodamico reviewed (approved):
Some comments/suggestions, but lgtm!

mariusandra committed:
Merge branch 'master' into persist-flags

mariusandra committed:
rename to PERSISTED feature flags

mariusandra commented:
Of course now that the flags are force-enabled, all cypress tests fail because we never developed test with the flags on. Probably the taxonomic filter is causing issues here 😅.

mariusandra committed:
remove flags

mariusandra commented:
On second thought, I will remove the released features themselves from this PR. Releasing them will require bigger changes to the tests and I fear this PR will get way too messy then.

Any objections to merging it in as is? I'll create separate a new PR for the actual flags.
